File: rust/day_1/part_1.rs
use std::fs;

fn solve(input: &str) -> i32 {
    let rotations: Vec<&str> = input
        .trim()
        .lines()
        .map(|line| line.trim())
        .filter(|line| !line.is_empty())
        .collect();

    let mut position = 50;
    let mut zeros = 0;

    for r in rotations {
        let direction = r.chars().next().unwrap();
        let distance: i32 = r[1..].parse().unwrap();

        if direction == 'L' {
            position = (position - distance).rem_euclid(100);
        } else {
            position = (position + distance).rem_euclid(100);
        }

        if position == 0 {
            zeros += 1;
        }
    }

    zeros
}

fn main() {
    let input = fs::read_to_string("../inputs/day_1.txt").expect("Failed to read input file");
    let result = solve(&input);
    println!("{}", result);
}

File: go/day_1/part_1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(input string) int {
	rotations := []string{}
	for _, line := range strings.Split(strings.TrimSpace(input), "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			rotations = append(rotations, line)
		}
	}

	position := 50
	zeros := 0

	for _, r := range rotations {
		direction := r[0]
		distance, _ := strconv.Atoi(r[1:])

		if direction == 'L' {
			position = (position - distance) % 100
			if position < 0 {
				position += 100
			}
		} else {
			position = (position + distance) % 100
		}

		if position == 0 {
			zeros++
		}
	}

	return zeros
}

func main() {
	input, err := os.ReadFile("../inputs/day_1.txt")
	if err != nil {
		panic(err)
	}
	result := solve(string(input))
	fmt.Println(result)
}
